library/ssl_cli.c

```diff
@@ -1385,99 +1614,199 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
 #endif /* MBEDTLS_SSL_ALPN */
 
 /*
- * Parse HelloVerifyRequest.  Only called after verifying the HS type.
+ *
+ * STATE HANDLING: ServerHello
+ *
+ */
+
+/*
+ * Overview
  */
+
+/* Main entry point; orchestrates the other functions */
+static int ssl_process_server_hello( mbedtls_ssl_context *ssl );
+
+/* Fetch and preprocess
+ * Returns a negative value on failure, and otherwise
+ * - SSL_SERVER_HELLO_COORDINATE_HELLO or
+ * - SSL_SERVER_HELLO_COORDINATE_VERIFY
+ * to indicate which message is expected and to be parsed next. */
+#define SSL_SERVER_HELLO_COORDINATE_HELLO  0
+#define SSL_SERVER_HELLO_COORDINATE_VERIFY 1
+static int ssl_process_server_hello_coordinate( mbedtls_ssl_context *ssl );
+
+/* Parse ServerHello */
+static int ssl_process_server_hello_parse( mbedtls_ssl_context *ssl,
+                                           const unsigned char* buf,
+                                           size_t buflen );
+static int ssl_process_server_hello_postprocess( mbedtls_ssl_context *ssl );
+/* Parse HelloVerify */
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
-static int ssl_parse_hello_verify_request( mbedtls_ssl_context *ssl )
+static int ssl_process_hello_verify_parse( mbedtls_ssl_context *ssl,
+                                           const unsigned char* buf,
+                                           size_t buflen );
+static int ssl_process_hello_verify_postprocess( mbedtls_ssl_context *ssl );
+#endif
+
+/*
+ * Implementation
+ */
+
+static int ssl_process_server_hello( mbedtls_ssl_context *ssl )
 {
-    const unsigned char *p = ssl->in_msg + mbedtls_ssl_hs_hdr_len( ssl );
-    int major_ver, minor_ver;
-    unsigned char cookie_len;
+    int ret = 0;
+    int msg_expect;
 
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> parse hello verify request" ) );
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> process server hello" ) );
 
-    /*
-     * struct {
-     *   ProtocolVersion server_version;
-     *   opaque cookie<0..2^8-1>;
-     * } HelloVerifyRequest;
+    /* Preprocessing step is no-op */
```

Comment:
Why is this comment here? I may just show my lack of knowledge, but why bother commenting that something is not performed in this particular case?

Reply:
This was among the first functions I modified, and back then I was very verbose about documenting the parts of the restructured function, even if it meant that some parts are not no-ops for the respective handshake state. 
With hindsight, I agree it's not helpful, so I will remove it.